gcs_util.go

```diff
@@ -1,44 +1,243 @@
+		accessToken := meta.client.(string)
+		gcsHeaders := map[string]string{
+			"Authorization": "Bearer " + accessToken,
+		}
+
+		req, err := http.NewRequest("HEAD", URL.String(), nil)
+		if err != nil {
+			return nil
+		}
+		for k, v := range gcsHeaders {
+			req.Header.Add(k, v)
+		}
+		client := &http.Client{}
+		resp, _ := client.Do(req)
+		if resp.StatusCode != http.StatusOK {
+			if resp.StatusCode == 403 || resp.StatusCode == 408 || resp.StatusCode == 429 || resp.StatusCode == 500 || resp.StatusCode == 503 {
+				meta.lastError = fmt.Errorf(resp.Status)
+				meta.resStatus = needRetry
+				return nil
+			}
+			if resp.StatusCode == 404 {
+				meta.resStatus = notFoundFile
+			} else if util.isTokenExpired(resp) {
+				meta.lastError = fmt.Errorf(resp.Status)
+				meta.resStatus = renewToken
+			}
+			meta.lastError = fmt.Errorf(resp.Status)
+			meta.resStatus = errStatus
+			return nil
+		}
+
+		digest := resp.Header.Get(gcsMetadataSfcDigest)
+		contentLength, _ := strconv.Atoi(resp.Header.Get("content-length"))
+		var encryptionMeta *encryptMetadata
+		if resp.Header.Get(gcsMetadataEncryptionDataProp) != "" {
+			var encryptData *encryptionData
+			json.Unmarshal([]byte(resp.Header.Get(gcsMetadataEncryptionDataProp)), encryptData)
+			if encryptData != nil {
+				encryptionMeta = &encryptMetadata{
+					key: encryptData.WrappedContentKey.EncryptionKey,
+					iv:  encryptData.ContentEncryptionIV,
+				}
+				if resp.Header.Get(gcsMetadataMatdescKey) != "" {
+					encryptionMeta.matdesc = resp.Header.Get(gcsMetadataMatdescKey)
+				}
+			}
+		}
+		meta.resStatus = uploaded
+		return &fileHeader{
+			digest:             digest,
+			contentLength:      int64(contentLength),
+			encryptionMetadata: encryptionMeta,
+		}
+	}
+	return nil
 }
 
 // cloudUtil implementation
 func (util *snowflakeGcsUtil) uploadFile(dataFile string, meta *fileMetadata, encryptMeta *encryptMetadata, maxConcurrency int, multiPartThreshold int64) error {
-	// TODO SNOW-294155
+	uploadURL := meta.presignedURL
+	var accessToken string
+
+	if uploadURL == nil {
+		uploadURL = util.generateFileURL(meta.stageInfo.Location, strings.TrimLeft(meta.dstFileName, "/"))
+		accessToken = meta.client.(string)
+	}
+
+	var contentEncoding string
+	if meta.dstCompressionType != nil {
+		contentEncoding = strings.ToLower(meta.dstCompressionType.name)
+	}
+
+	if contentEncoding == "gzip" {
+		contentEncoding = ""
+	}
+
+	gcsHeaders := make(map[string]string)
+	gcsHeaders[httpHeaderContentEncoding] = contentEncoding
+	gcsHeaders[gcsMetadataSfcDigest] = meta.sha256Digest
+	if accessToken != "" {
+		gcsHeaders["Authorization"] = "Bearer " + accessToken
+	}
+
+	if encryptMeta != nil {
+		encryptData := encryptionData{
+			"FullBlob",
+			contentKey{
+				"symmKey1",
+				encryptMeta.key,
+				"AES_CBC_256",
+			},
+			encryptionAgent{
+				"1.0",
+				"AES_CBC_256",
+			},
+			encryptMeta.iv,
+			keyMetadata{
+				"Java 5.3.0",
```

Comment:
Java?

Reply:
Yup! It's the value of the encryption library, shared across all drivers.